In [26]:
import pandas as pd
from typing import Dict, List, Tuple
import json
from tqdm.auto import tqdm
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor

In [91]:
parquet_file_abcd_conv = Path("../data/ABCD_tripfiles_conv.parquet")
parquet_file_mnop_conv = Path("../data/MNOP_tripfiles_conv.parquet")
parquet_file_zyxw_conv = Path("../data/ZYXW_tripfiles_conv.parquet")

parquet_file_abcd_conv_test = Path("../data/ABCD_tripfiles_conv_test.parquet")
parquet_file_mnop_conv_test = Path("../data/MNOP_tripfiles_conv_test.parquet")
parquet_file_zyxw_conv_test = Path("../data/ZYXW_tripfiles_conv_test.parquet")

In [66]:
ab_conv_test = pd.read_parquet(parquet_file_abcd_conv_test)
# mn_conv_test = pd.read_parquet(parquet_file_mnop_conv_test)
# zy_conv_test = pd.read_parquet(parquet_file_zyxw_conv_test)

In [67]:
ab_conv_test

id        creation_time airline_code  flight_number  \
0       33902381  2024-05-01 09:14:42           AB           2359   
1       33910427  2024-05-01 10:28:12           AB           2104   
2       33910380  2024-05-01 10:28:11           AB           2109   
3       33909790  2024-05-01 10:28:12           AB           2110   
4       33902706  2024-05-01 09:14:41           AB           2127   
...          ...                  ...          ...            ...   
709608  33875192  2024-05-01 04:07:01           AB           2372   
709609  33875261  2024-05-01 04:08:01           AB           2372   
709610  33875264  2024-05-01 04:08:01           AB           2372   
709611  33875306  2024-05-01 04:09:01           AB           2372   
709612  33875309  2024-05-01 04:09:01           AB           2372   

        flight_date departure_airport     user_name         action_name  \
0                 1               BLR  service-acco     ASMMsgProcessor   
1                 2               BLR  service-acco     ASMMsgProcessor   
2                 2               BLR  service-acco     ASMMsgProcessor   
3                 2               BOM  service-acco     ASMMsgProcessor   
4                 2               BOM  service-acco     ASMMsgProcessor   
...             ...               ...           ...                 ...   
709608           30               GAU  service-acco  StorePaxDataAction   
709609           30               GAU  service-acco  StorePaxDataAction   
709610           30               GAU  service-acco  StorePaxDataAction   
709611           30               GAU  service-acco  StorePaxDataAction   
709612           30               GAU  service-acco  StorePaxDataAction   

                                              header_line  \
0       2024-05-01 09:14:42,420 INFO  [3f326d134d9de45...   
1       2024-05-01 10:28:12,745 INFO  [de24be75947007f...   
2       2024-05-01 10:28:11,301 INFO  [df3a1cc7d574161...   
3       2024-05-01 10:28:12,189 INFO  [9e66dfb1ffe9048...   
4       2024-05-01 09:14:41,884 INFO  [829566f527b588f...   
...                                                   ...   
709608  2024-05-01 04:07:01,675 INFO  [90c6642a105b86f...   
709609  2024-05-01 04:08:01,279 INFO  [76f51797d6d0b42...   
709610  2024-05-01 04:08:01,281 INFO  [76f51797d6d0b42...   
709611  2024-05-01 04:09:01,465 INFO  [ef79e34fa812d85...   
709612  2024-05-01 04:09:01,468 INFO  [ef79e34fa812d85...   

                                            entry_details  \
0       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...   
1       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...   
2       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...   
3       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...   
4       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...   
...                                                   ...   
709608  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...   
709609  com.onesystem.lc2.paxactuals.dto.PaxDataDTO [ ...   
709610  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...   
709611  com.onesystem.lc2.paxactuals.dto.PaxDataDTO [ ...   
709612  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...   

       data_CreateLoadSheetAction  
0                            None  
1                            None  
2                            None  
3                            None  
4                            None  
...                           ...  
709608                       None  
709609                       None  
709610                       None  
709611                       None  
709612                       None  

[709613 rows x 11 columns]

In [92]:
ab_conv = pd.read_parquet(parquet_file_abcd_conv)
mn_conv = pd.read_parquet(parquet_file_mnop_conv)
zy_conv = pd.read_parquet(parquet_file_zyxw_conv)

In [99]:
zy_conv.sort_values(by="action_name").action_name.value_counts(sort=False)

action_name
ASMMsgProcessor                         33836
AcceptTransitDataAction                     4
AcceptTransitLoadAction                     2
AcceptTransitNotocAction                    4
AssignLCCAction                         42336
AssignLoadplanAction                     5661
AssignUnassignViewAction                  713
AutoLoadBulkAction                      11180
AutoLoadULDAction                          96
AutomaticNotificationAction             11712
CalculateWeightAndTrimAction           131682
CargoFinalAction                           68
CargoFinalActionTDM                     11722
ChangeFlightLegStateAction               5070
CheckinMsgProcessor                       260
ClearFlightsAction                      10846
CloseLegAction                              2
CopyCargoTableAction                        2
CopyLoadItemsAction                         2
CopyPaxDataAction                           4
CopyPaxLoadDataAction                       4
CreateAndSendFuelOrder

data_CalculateWeightAndTrimAction         131682
data_CheckinMsgProcessor                  130
data_CreateLoadingInstructionAction       43    X G
data_CreateLoadsheetAction                17    X G
data_CreateZFWMessageAction               5997
data_EstimateStorePaxDataAction           0     X
data_RampFinalAction                      38    X
data_SendFuelOrderAction                  3     X
data_SendLoadingInstructionAction         1     X
data_SendLoadsheetAction                  1     X
data_SetActualBagWeightIndicatorAction    5786
data_SetCKIPaxDistributionAction          5799
data_StoreAircraftDataAction              68    X
data_StorePaxDataAction                   9068
data_StorePaxDataGuiAction                21    X
data_StoreRegistrationAndConfigurationAc  9774
data_TdmCreateLoadingInstructionAction    2     X
data_TransferCargoAction                  12564
data_TransferCheckinDataAction            0     X G
data_UpdateEstimatesAction                0     X G
data_UpdateFuelDataAction                 125
data_UpdateLoadTableAction                4038
data_UpdateTransitLoadTableAction         4     X

data_EstimateStorePaxDataAction in MNOP ZYXW

In [98]:
zy_conv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 429759 entries, 0 to 429758
Data columns (total 33 columns):
 #   Column                                    Non-Null Count   Dtype 
---  ------                                    --------------   ----- 
 0   id                                        429759 non-null  int64 
 1   creation_time                             429759 non-null  object
 2   airline_code                              429759 non-null  object
 3   flight_number                             429759 non-null  int64 
 4   flight_date                               429759 non-null  int64 
 5   departure_airport                         429759 non-null  object
 6   user_name                                 429759 non-null  object
 7   action_name                               429759 non-null  object
 8   header_line                               429759 non-null  object
 9   entry_details                             384645 non-null  object
 10  data_CalculateWeightAndTrimActio

In [86]:
print(ab_conv[ab_conv.action_name == "SendLoadsheetAction"].entry_details.iloc[1])


Email receivers     :
Message type        :   LOADSHEET
Subject             :   AB2541/02MAY CCU-BLR; Loadsheet Edition 02
Attachment file name:   AB2541_02MAY_CCUBLR_Loadsheet_Edition_02
Body                :
 ABCD AIR
 L O A D S H E E T             CHECKED:         APPROVED     EDNO
 ALL WEIGHTS IN KILOGRAM       CLC 001                        02

 FROM/TO FLIGHT      A/C REG    CONFIG     CREW    DATE      TIME
 CCU BLR AB 2541/02  ABCAD      Y189       3/05   02MAY24    1721
 
                             WEIGHT DISTRIBUTION
 LOAD IN COMPARTMENTS         2120   1/0  2/780  3/1045  4/295

                             BAG/1340 CARGO/780 MAIL/0 EIC/0

 PASSENGER / CABIN BAG       12410  124/039/005/001     TTL 169
 CABIN CLASS                        Y  /168         
 CABIN AREA                          0A/25  0B/29  0C/26  0D/27  
                                     0E/30  0F/31    
 TOB                                177

 TOTAL TRAFFIC LOAD          14530
 DRY OPERATING WEIGHT    

In [15]:
weights_to_extract = {"RampFinalAction": {"keys": {"EZFW": "EZFW"}}}

In [16]:
class Weight:
    def __init__(self, key, description=None):
        self.key = key
        self.description = description

    def get_key(self):
        return self.key

    def get_description(self):
        return self.description

In [17]:
EZFW = Weight("EZFW", "Estimated Zero Fuel Weight")
AZFW = Weight("AZFW", "Actual Zero Fuel Weight")
ETOW = Weight("ETOW", "Estimated Takeoff Weight")
ATOW = Weight("ATOW", "Actual Takeoff Weight")

In [31]:
class Action:
    def __init__(self, name: str, weights: Dict[Weight, str]):
        self._name = name
        self._weights = weights

    def get_name(self):
        return self._name

    def get_weights(self):
        return self._weights

    def get_weight_items(
        self,
    ):
        return self._weights.items()

In [19]:
# RampFinalAction = Action(name="RampFinalAction", weights={EZFW: "EZFW"})

In [52]:
foo = [
    # Action(name="RampFinalAction", weights={EZFW: "EZFW"}), # EZFW is not a value just a status
    Action(name="CreateLoadsheetAction", weights={ATOW: "TAKE_OFF_WEIGHT_ACTUAL"}),
    Action(name="SendLoadsheetAction", weights={ATOW: "TAKE_OFF_WEIGHT_ACTUAL"}),
]

In [48]:
df = ab_conv.copy()

In [44]:
# Recursive function to find the value for a given key
def find_value(data: dict | list, key: str):
    if isinstance(data, dict):
        for k, v in data.items():
            if k == key:
                return v
            else:
                found = find_value(v, key)
                if found is not None:
                    return found
    elif isinstance(data, list):
        for item in data:
            found = find_value(item, key)
            if found is not None:
                return found
    return None


# Function to apply the recursive search to JSON data
def extract_key(json_str: str, key: str):
    data = json.loads(json_str)
    return find_value(data, key)

In [56]:
ab_conv.action_name.unique()

array(['ASMMsgProcessor', 'AcceptTransitDataAction',
       'AcceptTransitLoadAction', 'AcceptTransitNotocAction',
       'AssignLCCAction', 'AssignLoadplanAction',
       'AssignUnassignViewAction', 'CalculateWeightAndTrimAction',
       'CargoFinalAction', 'ChangeFlightLegStateAction',
       'CheckinMsgProcessor', 'CloseLegAction',
       'CreatePostDepartureMessagesAction',
       'CreateLoadingInstructionAction', 'CreateLoadsheetAction',
       'CreateZFWMessageAction', 'CreateBaggageLoadItemsAction',
       'ClearFlightsAction', 'InternalCreateLoadingInstructionAct',
       'InternalCreateLoadsheetAction', 'FlightPlanFiguresInMsgProcessor',
       'RampFinalAction', 'CrewMsgProcessor', 'PAXBOOKINGINMsgProcessor',
       'EstimateStorePaxDataAction', 'ReopenLegAction',
       'GetCabinConfigurationsAction', 'FuelDataInitializer',
       'SendPostDepartureMessagesAction',
       'SetActualBagWeightIndicatorAction', 'TransferCheckinDataAction',
       'StoreAircraftDataAction', 'Spe

In [50]:
df = df.replace({None: pd.NA})

In [53]:
for action in foo:
    for weight, key in action.get_weight_items():

        mask = (df.action_name == action.get_name()) & (
            ~df[f"data_{action.get_name()}"].isna()
        )

        df[weight.get_key()] = df[mask][f"data_{action.get_name()}"].apply(
            lambda x: extract_key(x, key)
        )

In [54]:
df

id        creation_time airline_code  flight_number  \
0       33902381  2024-05-01 09:14:42           AB           2359   
1       33910427  2024-05-01 10:28:12           AB           2104   
2       33910380  2024-05-01 10:28:11           AB           2109   
3       33909790  2024-05-01 10:28:12           AB           2110   
4       33902706  2024-05-01 09:14:41           AB           2127   
...          ...                  ...          ...            ...   
709608  33875192  2024-05-01 04:07:01           AB           2372   
709609  33875261  2024-05-01 04:08:01           AB           2372   
709610  33875264  2024-05-01 04:08:01           AB           2372   
709611  33875306  2024-05-01 04:09:01           AB           2372   
709612  33875309  2024-05-01 04:09:01           AB           2372   

        flight_date departure_airport     user_name         action_name  \
0                 1               BLR  service-acco     ASMMsgProcessor   
1                 2               BLR  service-acco     ASMMsgProcessor   
2                 2               BLR  service-acco     ASMMsgProcessor   
3                 2               BOM  service-acco     ASMMsgProcessor   
4                 2               BOM  service-acco     ASMMsgProcessor   
...             ...               ...           ...                 ...   
709608           30               GAU  service-acco  StorePaxDataAction   
709609           30               GAU  service-acco  StorePaxDataAction   
709610           30               GAU  service-acco  StorePaxDataAction   
709611           30               GAU  service-acco  StorePaxDataAction   
709612           30               GAU  service-acco  StorePaxDataAction   

                                              header_line  \
0       2024-05-01 09:14:42,420 INFO  [3f326d134d9de45...   
1       2024-05-01 10:28:12,745 INFO  [de24be75947007f...   
2       2024-05-01 10:28:11,301 INFO  [df3a1cc7d574161...   
3       2024-05-01 10:28:12,189 INFO  [9e66dfb1ffe9048...   
4       2024-05-01 09:14:41,884 INFO  [829566f527b588f...   
...                                                   ...   
709608  2024-05-01 04:07:01,675 INFO  [90c6642a105b86f...   
709609  2024-05-01 04:08:01,279 INFO  [76f51797d6d0b42...   
709610  2024-05-01 04:08:01,281 INFO  [76f51797d6d0b42...   
709611  2024-05-01 04:09:01,465 INFO  [ef79e34fa812d85...   
709612  2024-05-01 04:09:01,468 INFO  [ef79e34fa812d85...   

                                            entry_details  ...  \
0       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...  ...   
1       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...  ...   
2       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...  ...   
3       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...  ...   
4       <?xml version="1.0" encoding="UTF-8"?>\r\n<ns2...  ...   
...                                                   ...  ...   
709608  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...  ...   
709609  com.onesystem.lc2.paxactuals.dto.PaxDataDTO [ ...  ...   
709610  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...  ...   
709611  com.onesystem.lc2.paxactuals.dto.PaxDataDTO [ ...  ...   
709612  TOTAL Pax: 131    Y: 131  Jump: 0    StandBy: ...  ...   

       data_StorePaxDataGuiAction data_StoreRegistrationAndConfigurationAc  \
0                            <NA>                                     <NA>   
1                            <NA>                                     <NA>   
2                            <NA>                                     <NA>   
3                            <NA>                                     <NA>   
4                            <NA>                                     <NA>   
...                           ...                                      ...   
709608                       <NA>                                     <NA>   
709609                       <NA>                                     <NA>   
709610                       <NA>                       

In [22]:
# for name, keys in weights_to_extract.items():
#     print(name, keys)
#     for key, value in keys["keys"].items():
        